orttraining/orttraining/training_api/checkpoint.h

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if defined(ENABLE_TRAINING) && defined(ENABLE_TRAINING_ON_DEVICE)
+
+#pragma once
+#include "core/platform/path_lib.h"
+#include "core/platform/env.h"
+#include "orttraining/core/framework/checkpointing.h"
+#include "orttraining/training_api/interfaces.h"
+#include <type_traits>
+#include "onnx/defs/tensor_proto_util.h"
+
+namespace onnxruntime {
+namespace training {
+
+// TODO: Rename to api after all major classes implemented.
+namespace api_test {
+
+/**
+ * @brief Base class for user defined checkpoint property.
+ */
+struct CheckpointProperty {
+ public:
+  CheckpointProperty(const std::string& prop_name)
+      : prop_name_(prop_name) {
+  }
+
+  virtual ~CheckpointProperty() {}
+  virtual ONNX_NAMESPACE::TensorProto ToTensorProto() = 0;
+
+  std::string GetName() const {
+    return prop_name_;
+  }
+
+ protected:
+  std::string prop_name_;
+};
+
+/**
+ * @brief User defined checkpoint property.
+ * Only support int64_t, std::string and float data types.
+ */
+template <typename T>
+struct TypedCheckpointProperty : public CheckpointProperty {
+ public:
+  TypedCheckpointProperty(const std::string& prop_name, const T& prop_value)
+      : CheckpointProperty(prop_name), prop_value_(prop_value) {
+  }
+
+  ONNX_NAMESPACE::TensorProto ToTensorProto() override {
+    auto t_proto = ONNX_NAMESPACE::ToTensor<T>(prop_value_);
+
+    // We did not add dims assuming currently we only support scalars.
+    return t_proto;
+  }
+
+  T GetData() const {
+    return prop_value_;
+  }
+
+ private:
+  T prop_value_;
+};
+
+/**
+ * @brief A data class representing traing states.
+ * Including:
+ *     > Parameter states.
+ *     > Optimizer states.
+ *     > User defined training properties, for example 'epoch', 'best_score'.
+ */
+struct CheckpointStates {
+ public:
+  ModuleCheckpointStates module_checkpoint_states;
+  OptimizerCheckpointStates optimizer_checkpoint_states;
+  std::unordered_map<std::string, std::shared_ptr<CheckpointProperty>> named_properties;
+};
+
+PathString CreateFolderIfNotExists(const PathString& path, const std::string& folder_name);
+
+Status CreateOrtValuesFromTensorProtos(const PathString& model_location,
```

Comment:
why is this public? load should internally take care of this... Load should be able to load a tensorproto from the file into a paramter object